### PR TITLE
feat: add data struct for inference schema

### PR DESCRIPTION
**Pull request by khorshuheng**
**What this PR does / why we need it**:
This PR provides the data types required to support ML observability within Merlin. The new data types are expected to be used for https://github.com/caraml-dev/merlin/pull/488 , flyte workflows, and also in the future when we store an optional inference schema within Merlin.

dataclasses_json dependency is introduced here as it is required if we want to pass dataclass as an input to Flyte.

**Which issue(s) this PR fixes**:

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Checklist**

- [ ] Added unit test, integration, and/or e2e tests
- [ ] Tested locally
- [ ] Updated documentation
- [ ] Update Swagger spec if the PR introduce API changes
- [ ] Regenerated Golang and Python client if the PR introduce API changes

### CODE REVIEW

**Review by krithika369 — APPROVED** [review]
LGTM. Left some small comments. Thanks, @khorshuheng !

File: python/sdk/setup.py [hunk 1]
```diff
@@ -28,6 +28,7 @@
     "Click>=7.0,<8.1.4",
     "cloudpickle==2.0.0",  # used by mlflow
     "cookiecutter>=1.7.2",
+    "dataclasses-json>=0.6.0",
```
[the remark this answers is not in the document]
Comment (khorshuheng) [reply python/sdk/setup.py:31]: I can downgrade this to 0.5.2

File: python/sdk/merlin/observability/inference.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+class InferenceType(Enum):
+    BINARY_CLASSIFICATION = 1
+    MULTICLASS_CLASSIFICATION = 2
+    REGRESSION = 3
+    RANKING = 4
+
+
+@dataclass
+@dataclass_json
+class InferenceSchema:
```
[the remark this answers is not in the document]
Comment (khorshuheng) [reply python/sdk/merlin/observability/inference.py:82]: Originally, i plan to move ModelSpec from #488 to here as it is without any change. However, I realized that we might want to include the schema as part of the Merlin Model class attributes. If we do that, then it doesn't make sense to include model id and model version. I also renamed the class from ModelSpec to InferenceSchema.

File: python/sdk/merlin/observability/inference.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    @property
+    def column_types(self) -> Dict[str, ValueType]:
+        return {self.prediction_score_column: ValueType.FLOAT64}
+
+
+@dataclass_json
+@dataclass
+class BinaryClassificationOutput:
+    prediction_label_column: str
+    prediction_score_column: Optional[str] = None
```
Comment (tiopramayudi) [inline python/sdk/merlin/observability/inference.py:30]: can we make it mandatory and `prediction_label_column` is optional?

File: python/sdk/merlin/observability/inference.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if self.prediction_score_column is not None:
+            column_types_mapping[self.prediction_score_column] = ValueType.FLOAT64
+        return column_types_mapping
+
+
+@dataclass_json
+@dataclass
+class MulticlassClassificationOutput:
+    prediction_label_columns: List[str]
+    prediction_score_columns: Optional[List[str]] = None
```
Comment (tiopramayudi) [inline python/sdk/merlin/observability/inference.py:44]: can we make it mandatory and `prediction_label_columns` is optional?

File: python/sdk/merlin/observability/inference.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            for column_name in self.prediction_score_columns:
+                column_types_mapping[column_name] = ValueType.FLOAT64
+        return column_types_mapping
+
+
+@dataclass_json
+@dataclass
+class RankingOutput:
+    rank_column: str
+    prediction_group_id_column: str
```
Comment (tiopramayudi) [inline python/sdk/merlin/observability/inference.py:62]: you can add new field `prediction_score_column` as optional field

**Review by tiopramayudi — COMMENTED** [review]
should we add conversion to arize schema here?

File: python/sdk/merlin/observability/inference.py (hunk 3 above)
Comment (khorshuheng) [reply python/sdk/merlin/observability/inference.py:30]: Since this is merged already, i will make the changes in #488 instead.

Comment (khorshuheng) [reply python/sdk/merlin/observability/inference.py:30]: As for conversion to Arize schema - i am not sure if we should add dependency to Arize on the Merlin SDK level. My preference would be to keep the dependency to Arize on the consumer / flyte workflow instead.

Comment (tiopramayudi) [reply python/sdk/merlin/observability/inference.py:30]: one drawback is easy to overlook if there is something change on the merlin part

**Pull request merged**